ocaml/xapi/xapi_guest_agent.ml

```diff
@@ -97,6 +98,45 @@ let networks path (list: string -> string list) =
 		|> List.map (fun (path, prefix) -> find_all_ips path prefix)
 		|> List.concat
 
+(* This function is passed the "device/vif" node, a function it can use to
+ * find the directory listing of sub-nodes and a function to retrieve the value
+ * with the given path.
+ * If "state" of all VIFs are "4", the return value is true
+ * which means the network paths are optimized.
+ * Or else the return value is false.
+ *)
+let network_paths_optimized path (list: string -> string list) (lookup: string -> string option) =
+	try
+		let _ = List.find (fun vif_id ->
+			let vif_state = lookup (extend (extend path vif_id) "state") in
+			match vif_state with
+			| Some vifstate -> vifstate <> "4"
+			| None -> true
+		) (list path) in
+		false
+	with Not_found -> true
```

Comment:
I find this function quite hard to follow, and it took me a while to determine that is it probably correct. This was probably because of the double (triple?) negative (`false` if we have not found not a 4, and `true` in some other cases).
I would prefer a simple fold, such as this:
```
List.fold_left (fun result vif_id ->
    let vif_state = lookup (extend (extend path vif_id) "state") in
    result && (vif_state = Some "4")
) true (list path)
```